File: app/models/listing.py
from .db import db

class Listing(db.Model):
    __tablename__ = "listings"

    id = db.Column(db.Integer, primary_key=True)
    description = db.Column(db.Text, nullable=False)
    price = db.Column(db.Integer, nullable=False)
    size = db.Column(db.Integer, nullable=False)
    is_available = db.Column(db.Boolean)
    year_built = db.Column(db.Integer, nullable=False)
    bedrooms = db.Column(db.Integer, nullable=False)
    bathrooms = db.Column(db.Integer, nullable=False)
    parking = db.Column(db.Boolean)
    laundry = db.Column(db.Boolean)
    address = db.Column(db.String(50), nullable=False)
    country = db.Column(db.String(255), nullable=False)
    city = db.Column(db.String(30), nullable=False)
    state = db.Column(db.String(15), nullable=False)
    zipcode = db.Column(db.Integer, nullable=False)
    url1 = db.Column(db.Text, nullable=False)
    url2 = db.Column(db.Text, nullable=False)
    url3 = db.Column(db.Text, nullable=False)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'))
    agent_id = db.Column(db.Integer, db.ForeignKey('agents.id'))

    agent_of_listing = db.relationship('Agent', back_populates='listing_agent')
    owner_listing = db.relationship('User', back_populates='listings_owner')
    images_of_listing = db.relationship('Image', back_populates='listing_images')
    bookings_listing = db.relationship('Booking', back_populates='listing_booking')

    def to_dict(self):
        return {
            'id': self.id,
            'description': self.description,
            'price': self.price,
            'is_available': self.is_available,
            'year_built': self.year_built,
            'size': self.size,
            'bedrooms': self.bedrooms,
            'bathrooms': self.bathrooms,
            'parking': self.parking,
            'laundry': self.laundry,
            'address': self.address,
            'country': self.country,
            'city': self.city,
            'state': self.state,
            'zipcode': self.zipcode,
            'url1': self.url1,
            'url2': self.url2,
            'url3': self.url3,
            'user_id': self.user_id,
            # 'images_of_listing': [image_of_listing.to_dict() for image_of_listing in self.images_of_listing]
        }

File: react-app/src/store/listing.js
import rfdc from "rfdc"
const clone = rfdc()

const LOAD_ALL_LISTING = "/api/listings/LOAD_LISTINGS"
const LOAD_LISTING = "/api/listings/LOAD_LISTING"
const CREATE_LISTING = "/api/listings/CREATE_LISTING"
const EDIT_LISTING = "/api/listings/EDIT_LISTING"
const DELETE_LISTING = "/api/listings/DELETE_LISTING"

const allListings = listings => {
    return {
        type: LOAD_ALL_LISTING,
        listings
    }
}

const oneListing = listing => {
    return {
        type: LOAD_LISTING,
        listing
    }
}

const addListing = listing => {
    return {
        type: CREATE_LISTING,
        listing
    }
}

const editListing = listing => {
    return {
        type: EDIT_LISTING,
        listing
    }
}

const deleteListing = listing => {
    return {
        type: DELETE_LISTING,
        listing
    }
}

export const getListings = () => async dispatch => {
    const response = await fetch("/api/listings/")
    if(response.ok) {
        const listings = await response.json()
        dispatch(allListings(listings))
    }
}

export const getListing = id => async dispatch => {
    const response = await fetch(`/api/listings/${id}`)
    if(response.ok) {
        const listing = await response.json()
        dispatch(oneListing(listing))
    }
}

export const createListing = data => async dispatch => {
    const formData = new FormData()
    // formData.append("title", data.title)
    formData.append("description", data.description)
    formData.append("price", data.price)
    formData.append("size", data.size)
    formData.append("is_available", data.is_available)
    formData.append("year_built", data.year_built)
    formData.append("parking", data.parking)
    formData.append("laundry", data.laundry)
    formData.append("bedrooms", data.bedrooms)
    formData.append("bathrooms", data.bathrooms)
    formData.append("address", data.address)
    formData.append("country", data.country)
    formData.append("city", data.city)
    formData.append("state", data.state)
    formData.append("zipcode", data.zipcode)
    formData.append("url1", data.url1)
    formData.append("url2", data.url2)
    formData.append("url3", data.url3)
    formData.append("user_id", data.user_id);
    formData.append("agent_id", data.agent_id);
    // formData.append("images_of_listing", data.images_of_listing)
    // console.log("data in thunk", data)

    const response = await fetch("/api/listings/", {
        method: "POST",
        body: formData,
    });

    if(response.ok) {
        const listing = await response.json()
        dispatch(addListing(listing))
        return listing
    }
}

export const updateListing = data => async dispatch => {
    const formData = new FormData()
    // formData.append("title", data.title)
    formData.append("description", data.description)
    formData.append("price", data.price)
    formData.append("is_available", data.is_available)
    formData.append("url1", data.url1)
    formData.append("url2", data.url2)
    formData.append("url3", data.url3)
    // console.log("data", data)
    const response = await fetch(`/api/listings/${data.id}`, {
        method: "PUT",
        body: formData
    })

    if(response.ok) {
        const editedListing = await response.json()
        dispatch(editListing(editedListing))
    }
}
export const removeListing = id => async dispatch => {
    const response = await fetch(`/api/listings/${id}`, {
        method: "DELETE"
    })

    if(response.ok) {
        const listingId = await response.json()
        // console.log("listingid in thunk", listingId)
        dispatch(deleteListing(listingId))
    }
}

const initialState = {}

const listingReducer = (state = initialState, action) => {
    let newState = clone(state)
    switch(action.type) {
        case LOAD_ALL_LISTING:
            const listings = action.listings.listings
            listings.forEach(listing => {
                newState[listing.id] = listing
            })
            return newState
        case LOAD_LISTING:
            newState[action.listing.id] = action.listing
            return newState
        case CREATE_LISTING:
            newState[action.listing.id] = action.listing;
            return newState;
        case EDIT_LISTING:
            delete newState[action.listing.id]
            newState[action.listing.id] = action.listing;
            return newState;
        case DELETE_LISTING:
            delete newState[action.listing.id]
            return newState
        default:
            return state
    }
}

export default listingReducer
